File: plugins/lyy-dev-plugin/skills/p3c-reviewer/scripts/diff_scan.py
"""Git 分支差异扫描 + P3C 代码审查合并脚本

功能:
- 获取两个分支之间的变更文件
- 对变更文件执行 P3C 检查
- 转换输出格式为简化数组
- 支持优先级过滤（默认 2 级）
"""

import argparse
import glob
import json
import os
import subprocess
import sys
from typing import Any, Dict, List, Tuple

if sys.platform == "win32":
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
    sys.stderr.reconfigure(encoding="utf-8", errors="replace")

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
LIB_DIR = os.path.join(SCRIPT_DIR, "lib")

P3C_RULESETS = [
    "rulesets/java/ali-comment.xml",
    "rulesets/java/ali-concurrent.xml",
    "rulesets/java/ali-constant.xml",
    "rulesets/java/ali-exception.xml",
    "rulesets/java/ali-flowcontrol.xml",
    "rulesets/java/ali-naming.xml",
    "rulesets/java/ali-oop.xml",
    "rulesets/java/ali-orm.xml",
    "rulesets/java/ali-other.xml",
    "rulesets/java/ali-set.xml",
]

# Priority 到 BlockLevel 的映射
PRIORITY_TO_BLOCK_LEVEL = {
    1: "Critical",
    2: "Critical",
    3: "Major",
    4: "Minor",
    5: "Info",  # 可选
}


# ---------------------------------------------------------------------------
# Git 仓库验证（来自 git_diff_files.py）
# ---------------------------------------------------------------------------

def validate_git_repo(repo_path: str) -> str:
    """验证路径为有效 Git 仓库，返回绝对路径。

    Args:
        repo_path: 仓库路径

    Returns:
        仓库的绝对路径

    Raises:
        SystemExit: 路径不存在或非 Git 仓库
    """
    abs_path = os.path.abspath(repo_path)
    if not os.path.isdir(abs_path):
        raise SystemExit(f"路径不存在或不是目录: {abs_path}")
    git_dir = os.path.join(abs_path, ".git")
    if not os.path.exists(git_dir):
        raise SystemExit(f"不是有效的 Git 仓库（未找到 .git）: {abs_path}")
    return abs_path


def validate_branch_exists(repo_path: str, branch: str) -> None:
    """验证分支在仓库中存在。

    Args:
        repo_path: 仓库绝对路径
        branch: 分支名

    Raises:
        SystemExit: 分支不存在或 git 不可用
    """
    cmd = ["git", "-C", repo_path, "rev-parse", "--verify", branch]
    try:
        proc = subprocess.run(
            cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE
        )
    except FileNotFoundError:
        raise SystemExit("git 命令未找到，请确认已安装 Git 并加入 PATH")
    if proc.returncode != 0:
        stderr_str = proc.stderr.decode("gbk", errors="replace").strip()
        raise SystemExit(f"分支不存在: {branch}\n  {stderr_str}")


# ---------------------------------------------------------------------------
# Git diff 执行（来自 git_diff_files.py）
# ---------------------------------------------------------------------------

def get_diff_file_statuses(
    repo_path: str, source: str, target: str
) -> List[Tuple[str, str]]:
    """执行 git diff 获取变更文件及其状态。

    使用三点语法 target...source 获取 source 相对 target 引入的变更。

    Args:
        repo_path: 仓库绝对路径
        source: 源分支（feature 分支）
        target: 目标分支（main/master）

    Returns:
        [(status, relative_path), ...] 列表

    Raises:
        SystemExit: git 命令执行失败
    """
    cmd = [
        "git", "-C", repo_path,
        "diff", "--name-status", f"{target}...{source}",
    ]
    try:
        proc = subprocess.run(
            cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE
        )
    except FileNotFoundError:
        raise SystemExit("git 命令未找到，请确认已安装 Git 并加入 PATH")

    stderr_str = proc.stderr.decode("gbk", errors="replace").strip()
    if proc.returncode != 0:
        raise SystemExit(f"git diff 执行失败:\n  {stderr_str}")
    if stderr_str:
        for line in stderr_str.splitlines()[:5]:
            print(f"  [WARN] {line}", file=sys.stderr)

    stdout_str = proc.stdout.decode("utf-8", errors="replace")
    results: List[Tuple[str, str]] = []
    for line in stdout_str.strip().splitlines():
        if not line.strip():
            continue
        parts = line.split("\t", 1)
        if len(parts) == 2:
            results.append((parts[0].strip(), parts[1].strip()))
    return results


# ---------------------------------------------------------------------------
# 文件过滤与路径解析（来自 git_diff_files.py）
# ---------------------------------------------------------------------------

def filter_changed_files(file_statuses: List[Tuple[str, str]]) -> List[str]:
    """过滤出非删除状态的变更文件。

    Args:
        file_statuses: [(status, path), ...] 列表

    Returns:
        相对路径列表
    """
    return [
        path for status, path in file_statuses
        if not status.startswith("D")
    ]


def resolve_absolute_paths(
    repo_path: str, relative_paths: List[str]
) -> List[str]:
    """将相对路径转为绝对路径，跳过不存在的文件并发出警告。

    Args:
        repo_path: 仓库绝对路径
        relative_paths: 相对于仓库根目录的文件路径列表

    Returns:
        存在的文件的绝对路径列表
    """
    absolute_paths: List[str] = []
    for rel_path in relative_paths:
        abs_path = os.path.normpath(os.path.join(repo_path, rel_path))
        if os.path.isfile(abs_path):
            absolute_paths.append(abs_path)
        else:
            print(
                f"  [WARN] 文件在工作区中不存在（可能未切换到 source 分支）: {rel_path}",
                file=sys.stderr,
            )
    return absolute_paths


# ---------------------------------------------------------------------------
# classpath（来自 batch_scan_files.py）
# ---------------------------------------------------------------------------

def build_classpath() -> str:
    """构建 PMD 和 P3C 的 classpath。

    Returns:
        由分隔符连接的 JAR 文件路径字符串
    """
    jars = glob.glob(os.path.join(LIB_DIR, "*.jar"))
    if not jars:
        raise FileNotFoundError(f"lib 目录下未找到 jar 文件：{LIB_DIR}")
    return os.pathsep.join(jars)


# ---------------------------------------------------------------------------
# PMD 执行（来自 batch_scan_files.py）
# ---------------------------------------------------------------------------

def run_p3c_check(source_path: str) -> str:
    """对单个文件执行 PMD 检查。

    Args:
        source_path: 文件路径

    Returns:
        JSON 格式的检查报告字符串
    """
    classpath = build_classpath()
    cmd = [
        "java", "-cp", classpath,
        "net.sourceforge.pmd.PMD",
        "-d", source_path,
        "-R", ",".join(P3C_RULESETS),
        "-f", "json",
        "--encoding", "UTF-8",
    ]

    proc = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    stdout_str = proc.stdout.decode("utf-8", errors="replace")
    stderr_str = proc.stderr.decode("utf-8", errors="replace")

    if stderr_str:
        for line in stderr_str.strip().splitlines()[:5]:
            print(f"  [WARN] {line}", file=sys.stderr)

    return stdout_str


# ---------------------------------------------------------------------------
# JSON 合并（来自 batch_scan_files.py）
# ---------------------------------------------------------------------------

def merge_reports(json_contents: List[str]) -> Dict[str, Any]:
    """合并多个 JSON 报告为一个。

    Args:
        json_contents: JSON 报告字符串列表

    Returns:
        合并后的 JSON 对象
    """
    merged_files: List[Dict[str, Any]] = []
    seen_files: set = set()

    for json_content in json_contents:
        if not json_content.strip():
            continue

        try:
            data = json.loads(json_content)
            files = data.get("files", [])
            for file_entry in files:
                filename = file_entry.get("filename")
                if filename and filename not in seen_files:
                    seen_files.add(filename)
                    merged_files.append(file_entry)
        except json.JSONDecodeError:
            continue

    return {"files": merged_files}


# ---------------------------------------------------------------------------
# 优先级过滤（来自 batch_scan_files.py）
# ---------------------------------------------------------------------------

def filter_by_priority(
    data: Dict[str, Any], max_priority: int = 2
) -> Dict[str, Any]:
    """过滤 JSON 报告，仅保留指定优先级以下的违规项。

    Args:
        data: JSON 报告对象
        max_priority: 最大优先级（保留 priority <= max_priority 的项）

    Returns:
        过滤后的 JSON 对象
    """
    filtered_files: List[Dict[str, Any]] = []

    for file_entry in data.get("files", []):
        filename = file_entry.get("filename", "")
        violations = file_entry.get("violations", [])

        filtered_violations = [
            v for v in violations
            if v.get("priority", 5) <= max_priority
        ]

        if filtered_violations:
            filtered_files.append({
                "filename": filename,
                "violations": filtered_violations
            })

    return {"files": filtered_files}


# ---------------------------------------------------------------------------
# 格式转换（新增）
# ---------------------------------------------------------------------------

def transform_to_output_format(
    data: Dict[str, Any],
    repo_path: str
) -> List[Dict[str, Any]]:
    """将 PMD JSON 报告转换为简化的输出格式。

    Args:
        data: PMD JSON 报告对象
        repo_path: 仓库路径（用于将绝对路径转为相对路径）

    Returns:
        简化格式的违规项列表
    """
    results: List[Dict[str, Any]] = []

    for file_entry in data.get("files", []):
        filename = file_entry.get("filename", "")

        # 将绝对路径转为相对于仓库的路径
        try:
            rel_path = os.path.relpath(filename, repo_path)
        except ValueError:
            rel_path = filename

        for violation in file_entry.get("violations", []):
            priority = violation.get("priority", 5)
            results.append({
                "fileName": rel_path,
                "beginline": violation.get("beginline", 0),
                "begincolumn": violation.get("begincolumn", 0),
                "endline": violation.get("endline", 0),
                "endcolumn": violation.get("endcolumn", 0),
                "suggestion": violation.get("description", ""),
                "ruleId": violation.get("rule", ""),
                "blockLevel": PRIORITY_TO_BLOCK_LEVEL.get(priority, "Info"),
            })

    return results


# ---------------------------------------------------------------------------
# 主流程
# ---------------------------------------------------------------------------

def diff_scan(
    repo_path: str,
    source: str,
    target: str,
    max_priority: int = 2
) -> List[Dict[str, Any]]:
    """执行分支差异扫描。

    Args:
        repo_path: Git 仓库路径
        source: 源分支
        target: 目标分支
        max_priority: 最大优先级（过滤阈值）

    Returns:
        简化格式的违规项列表
    """
    # 1. 验证 Git 仓库
    repo_abs_path = validate_git_repo(repo_path)
    validate_branch_exists(repo_abs_path, source)
    validate_branch_exists(repo_abs_path, target)

    # 2. 获取变更文件
    file_statuses = get_diff_file_statuses(repo_abs_path, source, target)
    changed_files = filter_changed_files(file_statuses)
    absolute_paths = resolve_absolute_paths(repo_abs_path, changed_files)

    if not absolute_paths:
        return []

    # 3. 执行 P3C 检查
    json_contents = []
    for file_path in absolute_paths:
        print(f"  [扫描] {os.path.basename(file_path)} ...", file=sys.stderr)
        json_content = run_p3c_check(file_path)
        json_contents.append(json_content)

    # 4. 合并并过滤报告
    merged_data = merge_reports(json_contents)
    filtered_data = filter_by_priority(merged_data, max_priority)

    # 5. 转换输出格式
    return transform_to_output_format(filtered_data, repo_abs_path)


# ---------------------------------------------------------------------------
# CLI 接口
# ---------------------------------------------------------------------------

def build_argument_parser() -> argparse.ArgumentParser:
    """构建命令行参数解析器。"""
    parser = argparse.ArgumentParser(
        description="Git 分支差异扫描 + P3C 代码审查",
    )
    parser.add_argument(
        "repo",
        help="Git 仓库根目录路径",
    )
    parser.add_argument(
        "--source",
        required=True,
        help="源分支（feature 分支）",
    )
    parser.add_argument(
        "--target",
        default="origin/master",
        help="目标分支（默认 origin/master）",
    )
    parser.add_argument(
        "--priority",
        type=int,
        default=2,
        choices=[1, 2, 3, 4, 5],
        help="最大优先级，默认 2（过滤掉 2 级以上的违规）",
    )
    return parser


def main() -> None:
    """主入口函数。"""
    parser = build_argument_parser()
    args = parser.parse_args()

    results = diff_scan(args.repo, args.source, args.target, args.priority)

    if not results:
        print("未发现违规项。")
        return

    print(json.dumps(results, indent=2, ensure_ascii=False))
    print(f"\n总违规数：{len(results)}", file=sys.stderr)


if __name__ == "__main__":
    main()

File: plugins/code-review/skills/java-code-review/scripts/diff_scan.mjs
/**
 * Git 分支差异扫描 + P3C 代码审查合并脚本
 *
 * 功能:
 * - 获取两个分支之间的变更文件
 * - 对变更文件执行 P3C 检查
 * - 转换输出格式为简化数组
 * - 支持优先级过滤（默认 2 级）
 *
 * 用法:
 *   node diff_scan.mjs <repo> --source <branch> [--target <branch>] [--priority <1-5>] [-v]
 */

import { spawnSync } from "node:child_process";
import {
  existsSync,
  mkdirSync,
  mkdtempSync,
  readdirSync,
  rmSync,
  statSync,
  writeFileSync,
} from "node:fs";
import { dirname, isAbsolute, join, normalize, relative, resolve, sep } from "node:path";
import { tmpdir } from "node:os";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const SCRIPT_DIR = __dirname;
const LIB_DIR = join(SCRIPT_DIR, "lib");

let VERBOSE = false;

function log(msg) {
  if (VERBOSE) console.error(`DEBUG: ${msg}`);
}
function warn(msg) {
  console.error(`WARNING: ${msg}`);
}
function fatal(msg) {
  console.error(msg);
  process.exit(1);
}

// ---------------------------------------------------------------------------
// 常量
// ---------------------------------------------------------------------------

const P3C_RULESETS = [
  "rulesets/java/ali-comment.xml",
  "rulesets/java/ali-concurrent.xml",
  "rulesets/java/ali-constant.xml",
  "rulesets/java/ali-exception.xml",
  "rulesets/java/ali-flowcontrol.xml",
  "rulesets/java/ali-naming.xml",
  "rulesets/java/ali-oop.xml",
  "rulesets/java/ali-orm.xml",
  "rulesets/java/ali-other.xml",
  "rulesets/java/ali-set.xml",
];

// Priority 到 BlockLevel 的映射
const PRIORITY_TO_BLOCK_LEVEL = {
  1: "Critical",
  2: "Critical",
  3: "Major",
  4: "Minor",
  5: "Info",
};

// ---------------------------------------------------------------------------
// 参数解析
// ---------------------------------------------------------------------------

function parseArgs(argv) {
  let repo = null;
  let source = null;
  let target = "origin/master";
  let priority = 2;
  let verbose = false;

  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === "--source" && i + 1 < argv.length) {
      source = argv[++i];
    } else if (arg === "--target" && i + 1 < argv.length) {
      target = argv[++i];
    } else if (arg === "--priority" && i + 1 < argv.length) {
      priority = parseInt(argv[++i], 10);
      if (isNaN(priority) || priority < 1 || priority > 5) {
        fatal("--priority 必须为 1-5 之间的整数");
      }
    } else if (arg === "--verbose" || arg === "-v") {
      verbose = true;
    } else if (!arg.startsWith("--") && !repo) {
      repo = arg;
    }
  }

  if (!repo) fatal("错误: 缺少必需参数 <repo>");
  if (!source) fatal("错误: 缺少必需参数 --source");

  return { repo, source, target, priority, verbose };
}

// ---------------------------------------------------------------------------
// 执行子进程的辅助函数
// ---------------------------------------------------------------------------

function runGit(args, options = {}) {
  const result = spawnSync("git", args, {
    maxBuffer: 50 * 1024 * 1024,
    ...options,
  });
  return result;
}

// ---------------------------------------------------------------------------
// 环境前置检查
// ---------------------------------------------------------------------------

function validateJavaAvailable() {
  const result = spawnSync("java", ["-version"], { stdio: "pipe" });
  if (result.error) {
    fatal("java 命令未找到，请确认已安装 JDK/JRE 并加入 PATH");
  }
  if (result.status !== 0) {
    fatal("java -version 返回非零退出码，请检查 Java 安装");
  }
}

// ---------------------------------------------------------------------------
// Git 仓库验证
// ---------------------------------------------------------------------------

function validateGitRepo(repoPath) {
  const absPath = isAbsolute(repoPath) ? repoPath : resolve(repoPath);
  try {
    const stat = statSync(absPath);
    if (!stat.isDirectory()) fatal(`路径不是目录: ${absPath}`);
  } catch {
    fatal(`路径不存在: ${absPath}`);
  }

  const result = runGit(["-C", absPath, "rev-parse", "--git-dir"], {
    stdio: "pipe",
  });
  if (result.status !== 0) {
    fatal(`不是有效的 Git 仓库: ${absPath}`);
  }
  return absPath;
}

function validateBranchExists(repoPath, branch) {
  const result = runGit(
    ["-C", repoPath, "rev-parse", "--verify", branch],
    { stdio: "pipe" }
  );
  if (result.status !== 0) {
    const stderr = result.stderr ? result.stderr.toString().trim() : "";
    fatal(`分支不存在: ${branch}\n  ${stderr}`);
  }
}

// ---------------------------------------------------------------------------
// Git diff 执行
// ---------------------------------------------------------------------------

function getDiffFileStatuses(repoPath, source, target) {
  const result = runGit(
    ["-C", repoPath, "diff", "--name-status", `${target}...${source}`],
    { stdio: "pipe" }
  );

  const stderr = result.stderr ? result.stderr.toString().trim() : "";
  if (result.status !== 0) fatal(`git diff 执行失败:\n  ${stderr}`);
  if (stderr) {
    for (const line of stderr.split("\n").slice(0, 5)) {
      warn(`git diff stderr: ${line}`);
    }
  }

  const stdout = result.stdout ? result.stdout.toString("utf-8") : "";
  const results = [];
  for (const line of stdout.trim().split("\n")) {
    if (!line.trim()) continue;
    const parts = line.split("\t");
    if (parts.length === 3) {
      // R(ename) / C(opy): status\told_path\tnew_path → 取新路径
      results.push([parts[0].trim(), parts[2].trim()]);
    } else if (parts.length === 2) {
      results.push([parts[0].trim(), parts[1].trim()]);
    }
  }
  return results;
}

// ---------------------------------------------------------------------------
// 变更行范围提取
// ---------------------------------------------------------------------------

function getChangedLineRanges(repoPath, source, target) {
  const result = runGit(
    ["-C", repoPath, "diff", "-U0", `${target}...${source}`],
    { stdio: "pipe" }
  );

  if (result.status !== 0) {
    const stderr = result.stderr ? result.stderr.toString().trim() : "";
    fatal(`git diff -U0 执行失败:\n  ${stderr}`);
  }

  const stdout = result.stdout ? result.stdout.toString("utf-8") : "";
  const changedRanges = {};
  let currentFile = null;

  for (const line of stdout.split("\n")) {
    if (line.startsWith("+++ b/")) {
      currentFile = line.slice(6);
    } else if (line.startsWith("@@") && currentFile) {
      const match = line.match(/\+(\d+)(?:,(\d+))?/);
      if (match) {
        const start = parseInt(match[1], 10);
        const count = match[2] !== undefined ? parseInt(match[2], 10) : 1;
        if (count > 0) {
          if (!changedRanges[currentFile]) changedRanges[currentFile] = [];
          changedRanges[currentFile].push([start, start + count - 1]);
        }
      }
    }
  }

  return changedRanges;
}

// ---------------------------------------------------------------------------
// 文件过滤与路径解析
// ---------------------------------------------------------------------------

function filterChangedFiles(fileStatuses) {
  return fileStatuses
    .filter(([status]) => !status.startsWith("D"))
    .map(([, path]) => path);
}

function isTestFile(filePath) {
  const normalized = filePath.replace(/\\/g, "/");
  return normalized.includes("/src/test/") || normalized.startsWith("src/test/");
}

function getCurrentBranch(repoPath) {
  const result = runGit(
    ["-C", repoPath, "rev-parse", "--abbrev-ref", "HEAD"],
    { stdio: "pipe" }
  );
  if (result.status !== 0) return "";
  const branch = result.stdout.toString("utf-8").trim();
  return branch === "HEAD" ? "" : branch;
}

function resolveWorkingTreePaths(repoPath, relativePaths) {
  const absolutePaths = [];
  for (const relPath of relativePaths) {
    const absPath = normalize(join(repoPath, relPath));
    try {
      const stat = statSync(absPath);
      if (stat.isFile()) {
        absolutePaths.push(absPath);
      } else {
        warn(`不是文件: ${relPath}`);
      }
    } catch {
      warn(`文件在工作区中不存在: ${relPath}`);
    }
  }
  return absolutePaths;
}

function extractFilesFromBranch(repoPath, branch, relativePaths, destDir) {
  const extracted = [];
  for (const relPath of relativePaths) {
    const destPath = join(destDir, relPath);
    mkdirSync(dirname(destPath), { recursive: true });

    const result = runGit(
      ["-C", repoPath, "show", `${branch}:${relPath}`],
      { stdio: "pipe" }
    );

    if (result.status !== 0) {
      warn(`无法从分支 ${branch} 提取文件: ${relPath}`);
      continue;
    }

    writeFileSync(destPath, result.stdout);
    extracted.push(destPath);
  }
  return extracted;
}

// ---------------------------------------------------------------------------
// classpath
// ---------------------------------------------------------------------------

function buildClasspath() {
  if (!existsSync(LIB_DIR)) {
    fatal(`lib 目录不存在: ${LIB_DIR}`);
  }
  const jars = readdirSync(LIB_DIR)
    .filter((f) => f.endsWith(".jar"))
    .map((f) => join(LIB_DIR, f));

  if (jars.length === 0) {
    fatal(`lib 目录下未找到 jar 文件: ${LIB_DIR}`);
  }
  return jars.join(sep === "\\" ? ";" : ":");
}

// ---------------------------------------------------------------------------
// PMD 执行
// ---------------------------------------------------------------------------

function runP3cCheck(sourcePaths, classpath) {
  const args = [
    "-Dfile.encoding=UTF-8",
    "-cp",
    classpath,
    "net.sourceforge.pmd.PMD",
    "-d",
    sourcePaths.join(","),
    "-R",
    P3C_RULESETS.join(","),
    "-f",
    "json",
    "--encoding",
    "UTF-8",
  ];

  const result = spawnSync("java", args, {
    stdio: "pipe",
    maxBuffer: 50 * 1024 * 1024,
  });

  const stdout = result.stdout ? result.stdout.toString("utf-8") : "";
  const stderr = result.stderr ? result.stderr.toString("utf-8") : "";

  if (stderr) {
    for (const line of stderr.trim().split("\n").slice(0, 5)) {
      warn(`PMD stderr: ${line}`);
    }
  }

  // PMD 退出码约定: 0=无违规, 4=有违规, 其他=执行错误
  if (result.status !== 0 && result.status !== 4) {
    fatal(
      `PMD 执行失败 (exit code ${result.status}):\n  ${stderr.slice(0, 500)}`
    );
  }

  return stdout;
}

// ---------------------------------------------------------------------------
// 优先级过滤
// ---------------------------------------------------------------------------

function filterByPriority(data, maxPriority = 2) {
  const filteredFiles = [];
  for (const fileEntry of data.files || []) {
    const violations = (fileEntry.violations || []).filter(
      (v) => (v.priority ?? 5) <= maxPriority
    );
    if (violations.length > 0) {
      filteredFiles.push({ filename: fileEntry.filename || "", violations });
    }
  }
  return { files: filteredFiles };
}

// ---------------------------------------------------------------------------
// 格式转换
// ---------------------------------------------------------------------------

function transformToOutputFormat(data, repoPath) {
  const results = [];
  for (const fileEntry of data.files || []) {
    const filename = fileEntry.filename || "";
    let relPath;
    try {
      relPath = relative(repoPath, filename);
    } catch {
      relPath = filename;
    }

    for (const v of fileEntry.violations || []) {
      const priority = v.priority ?? 5;
      results.push({
        fileName: relPath,
        beginline: v.beginline ?? 0,
        begincolumn: v.begincolumn ?? 0,
        endline: v.endline ?? 0,
        endcolumn: v.endcolumn ?? 0,
        suggestion: v.description ?? "",
        ruleId: v.rule ?? "",
        blockLevel: PRIORITY_TO_BLOCK_LEVEL[priority] || "Info",
      });
    }
  }
  return results;
}

// ---------------------------------------------------------------------------
// 变更行过滤
// ---------------------------------------------------------------------------

function filterByChangedLines(results, changedRanges) {
  const filtered = [];
  for (const item of results) {
    const fileName = item.fileName.replace(/\\/g, "/");
    const ranges = changedRanges[fileName];
    if (!ranges) continue;

    const begin = item.beginline || 0;
    for (const [rStart, rEnd] of ranges) {
      if (rStart <= begin && begin <= rEnd) {
        filtered.push(item);
        break;
      }
    }
  }
  return filtered;
}

// ---------------------------------------------------------------------------
// 主流程
// ---------------------------------------------------------------------------

function diffScan(repoPath, source, target, maxPriority = 2) {
  // 1. 环境与仓库验证
  validateJavaAvailable();
  const repoAbsPath = validateGitRepo(repoPath);
  validateBranchExists(repoAbsPath, source);
  validateBranchExists(repoAbsPath, target);

  // 2. 获取变更文件与变更行范围
  const fileStatuses = getDiffFileStatuses(repoAbsPath, source, target);
  const changedRanges = getChangedLineRanges(repoAbsPath, source, target);
  let changedFiles = filterChangedFiles(fileStatuses);
  changedFiles = changedFiles.filter(
    (f) => f.endsWith(".java") && !isTestFile(f)
  );

  if (changedFiles.length === 0) return [];

  // 3. 获取文件并扫描
  const classpath = buildClasspath();
  const currentBranch = getCurrentBranch(repoAbsPath);
  const useWorktree = currentBranch === source;

  let jsonContent;
  let basePath;

  if (useWorktree) {
    // source 就是当前分支，直接读工作区文件
    const absolutePaths = resolveWorkingTreePaths(repoAbsPath, changedFiles);
    if (absolutePaths.length === 0) return [];
    log(`共 ${absolutePaths.length} 个文件待扫描（工作区）`);
    jsonContent = runP3cCheck(absolutePaths, classpath);
    basePath = repoAbsPath;
  } else {
    // source 非当前分支，提取到临时目录
    const tmpDir = mkdtempSync(join(tmpdir(), "p3c_scan_"));
    try {
      const absolutePaths = extractFilesFromBranch(
        repoAbsPath,
        source,
        changedFiles,
        tmpDir
      );
      if (absolutePaths.length === 0) return [];
      log(`共 ${absolutePaths.length} 个文件待扫描（临时目录）`);
      jsonContent = runP3cCheck(absolutePaths, classpath);
      basePath = tmpDir;
    } finally {
      rmSync(tmpDir, { recursive: true, force: true });
    }
  }

  // 4. 解析并过滤报告
  let reportData;
  try {
    reportData = jsonContent.trim() ? JSON.parse(jsonContent) : { files: [] };
  } catch {
    warn("PMD 输出无法解析为 JSON");
    reportData = { files: [] };
  }

  const filteredData = filterByPriority(reportData, maxPriority);

  // 5. 转换输出格式并过滤到变更行范围
  const results = transformToOutputFormat(filteredData, basePath);
  return filterByChangedLines(results, changedRanges);
}

// ---------------------------------------------------------------------------
// CLI 入口
// ---------------------------------------------------------------------------

const args = parseArgs(process.argv.slice(2));
VERBOSE = args.verbose;

const results = diffScan(args.repo, args.source, args.target, args.priority);

if (results.length === 0) {
  console.log("未发现违规项。");
} else {
  console.log(JSON.stringify(results, null, 2));
  log(`总违规数：${results.length}`);
}
